### ISSUE TITLE
Feature request: Add more programming-like functions?

### ISSUE THREAD

**Issue by user202729** [typst/typst#244]
E.g.

* write to file
* call shell process
* communicate through pipe/socket?

My main motivation is to allow e.g. executing Python code from inside the document.

Unfortunately this would slow down incremental compilation a bit similar to #93. Still, being allowed to incorporate outputs of other programs (such as what `pythontex` or `sagetex` LaTeX package does), without having to manually save the result as a png/pdf and include it in Typst would be powerful.

----

Remark, it may be a security hole, so maybe only enable it if some flag is passed in similar to `--shell-escape` in TeX. See also discussion in #219.

**Comment by shizhaojingszj** [comment]
Cannot agree more, there are some similar projects like RMarkdown and Quarto, both from RStudio.

Quarto: https://github.com/quarto-dev/quarto-cli

**Comment by Enivex** [comment]
This seems like it would go against their core design principles

**Comment by user202729** [comment]
Indeed, there's a little problem here...

> In Typst, all functions are pure. This means that for the same arguments, they always return the same result. They cannot "remember" things to produce another value when they are called a second time.

and

> Performance through Incrementality: All Typst language features must accommodate for incremental compilation. Luckily we have [comemo](https://github.com/typst/comemo/), a system for incremental compilation which does most of the hard work in the background.

Maybe I'll propose the following as a compromise: have a function `readshell(cmd, input)` that runs the command `cmd` externally, passing `input` to its stdin, then read its stdout and interpret it as Typst code.

If the user ensure `cmd` is pure, then the function is still pure and memoization is possible.

**Comment by Dherse** [comment]
A bit of a SPAM as I already posted this exact answer on another issue, but I feel like re-iterating:

I am just here to comment regarding the idea of a `--shell-escape` flag. (please don't). There are three main reasons why that is not desirable: (note that the first two also apply to writing to file and sockets/pipes)

1. it would be a giant security hole (especially once the package manager arrives)
2. it would lead to non-deterministic rendering (different version of **external** CLI tools producing different outputs)
3. it would lead to a fracturing of the ecosystem between windows, linux and macos (I can't wait to have to do `if platform() == "windows"` in my **text** document)

Additionally, as previously mentioned [here](https://github.com/typst/typst/issues/295), Typst does not **need** to be a generic language, if features that cannot be created using the default language exist (which they probably do), I believe that there should be a plugin API (most likely using WASM as it is a decent encapsulation of third party code with near-native performance). It fixes all of the aforementioned issues.

Additionally, looking at the tectonic project, they did not want to add a `--shell-escape` flag until the community was really bullying them into doing so. I think that the LaTeX ecosystem would be better without this feature, and I strongly believe that, in the long term, this will be better for the Typst ecosystem as well.

**Comment by patrick-kidger** [comment]
I've written a package [typst_pyimage](https://github.com/patrick-kidger/typst_pyimage) that does exactly this: runs inline Python code to generate images.

Example:
```typst
#import ".typst_pyimage/pyimage.typ": pyimage

Consider the Lotka--Volterra (predator-prey)
equations:

#pyimage("
import diffrax
import jax.numpy as jnp
import matplotlib.pyplot as plt

def func(t, y, args):
  rabbits, cats = y
  d_rabbits = rabbits - rabbits*cats
  d_cats = -cats + rabbits*cats
  return d_rabbits, d_cats

term = diffrax.ODETerm(func)
solver = diffrax.Tsit5()
y0 = (2, 1)
t0 = 0
t1 = 20
dt0 = 0.01
ts = jnp.linspace(t0, t1, 100)
saveat = diffrax.SaveAt(ts=ts)
sol = diffrax.diffeqsolve(term, solver, t0,
                          t1, dt0, y0,
                          saveat=saveat)

plt.plot(ts, sol.ys[0], label='Rabbits')
plt.plot(ts, sol.ys[1], label='Cats')
plt.xlim(0, 20)
plt.ylim(0, 2.5)
plt.xlabel('Time')
plt.ylabel('Population')
plt.legend()
", width: 70%)
```

This makes it much easier to insert generated figures into my documents!

**Comment by user202729** [comment]
Interesting plugin. But the limitations in this plugin (use regex to extract Python blocks, need an external script to compile) are precisely the limitations caused by Typst not having an ability to e.g. write to a text file.

Which also mean that if there are two such plugins (that needs external script) they will not be usable together.

I think think of a few reasonable compromises:

* allow some sort of "write to standard output/designated file" feature. If PDF is a possible output format, why shouldn't TXT be?

  Because functions must be immutable however, maybe make it so that a [content] element will write whatever needed on rendering.

* add some argument types like `--run-cmd-before-compile` and `--run-cmd-after-compile` to Typst compiler. That way we can chain multiple such "packages".


------

Side note for this particular package, maybe accept backtick-delimited "strings" as input too? Refer to #266 .

**Comment by patrick-kidger** [comment]
Yeah, the limitations you discuss are all real. It'd be great to have a way for Typst to call out to an external program. (Assumed to be functionally pure.)

Thanks for the thoughts on typst_pyimage in particular. The file names are an internal implementation detail (they're placed inside the `.typst_pyimage` folder) but the rest SGTM.

**Comment by ruifengx** [comment]
I use [lhs2TeX](https://github.com/kosmikus/lhs2tex) to typeset Haskell code (and the code blocks are rendered like maths formulae instead of in monospace font). The tool works by first tokenizing and parsing the input, and I think most people would agree that the Typst language is not the best language to implement a Haskell parser. For Typst to support such use cases, it is necessary to have support for either plug-ins or running command-line programs, and the plug-in system should ideally support linking to languages other than Rust. From this point of view, WASM plug-ins are inferior to native plug-ins, because the former cannot link to a Haskell/C library (it may still do the job by delegate to a command-line program, if plug-ins are allowed to spawn subprocesses).

Personally, I do not think we will have a plug-in system soon (because (1) it apparently requires careful design, and (2) plug-ins tend to couple tightly with the Typst API, which will probably remain unstable for quite a while). Therefore, I am in favour of having a quick and dirty solution for now (i.e., to allow spawning subprocesses and collect Typst code from its stdout). It can be deprecated and removed later if we have a proper solution, but I hope we get a workaround while waiting for _the correct solution™_ to land.

Regarding the arguments against running command-line programs:
> 1. It would be a giant security hole (especially once the package manager arrives)

Of course it is, and the same applies to every program we install. In general, I do not read every line of code in every software before using them; I usually just trust the author. Yes, the `--shell-escape` option is insecure, but if it is the only option for now, we still need to get the work done.

> 2. It would lead to non-deterministic rendering (different version of **external** CLI tools producing different outputs)

IMO non-deterministic rendering is better than no rendering. Besides, the rendering is already non-deterministic in this sense because different versions of dependency `.typ` file would produce different outputs. In the end, both could be solved by proper version checking.

> 3. It would lead to a fracturing of the ecosystem between Windows, Linux and MacOS (I can't wait to have to do `if platform() == "windows"` in my **text** document)

I would argue that the fracturing only happens where the typesetting enabled by said package is impossible or extremely hard to implement using what Typst currently provides. Also, this will not be a problem if the dependency program is actually available on Windows, Linux, and MacOS.

In conclusion, I see no problem in having the ability to run external programs (as a temporary workaround, before a proper solution is available), especially if this feature is opt-in.

**Comment by Enivex** [comment]
> long rant

typst already has syntax highlighting built in https://typst.app/docs/reference/text/raw/

**Comment by ruifengx** [comment]
In my previous comment, I intended to describe another use case and illustrate why (IMO) typesetting it necessarily requires external programs/libraries/plug-ins. I'd like to make it clear that I know `raw` and it does not typeset the code in the desired way. The problem is that `raw` in its current form is not easily extensible: it does not use `tree-sitter` (#967), and it cannot render the code as maths formulae ([lhs2TeX](https://github.com/kosmikus/lhs2tex)).

My apologies for going too far beyond describing the use case.

**Comment by user202729** [comment]
Closing this in favor of #1056.
